ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreCleanUpTask.java

```diff
@@ -0,0 +1,199 @@
+          });
+
+      metrics.propertyStoreTaskTimeInMs.update(System.currentTimeMillis() - startTimeMs);
+      return new TaskResult(TaskResult.Status.COMPLETED, "PropertyStoreCleanUpTask completed successfully");
+    } catch (Exception exception) {
+      logger.error("Exception thrown while executing PropertyStoreCleanUpTask for cluster = {}"
+          ,manager.getClusterName(), exception);
+      metrics.propertyStoreCleanUpErrorCount.inc();
+      return new TaskResult(TaskResult.Status.FAILED, "Exception thrown");
+    }
+  }
+
+  /**
+   *  Cleanup should be done if the instance is not live and not present in ideal state or external view
+   * @param liveInstances Set of live instances
+   * @param instancesInIdealStateAndExternalView Set of instances in ideal state and external view
+   * @return
+   */
+  private boolean shouldDoCleanUpPropertyStore(String instance, Set<String> liveInstances,
+      Set<String> instancesInIdealStateAndExternalView) {
+    if(liveInstances.contains(instance)) {
+      logger.trace("Instance {} is live, skipping cleanup", instance);
+      return false;
+    }
+    if(instancesInIdealStateAndExternalView.contains(instance)) {
+      logger.trace("Instance {} is present in ideal state or external view, skipping cleanup", instance);
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Cleanup property store for the given instance
+   */
+  private void cleanupPropertyStore(String instance) {
+    if(clusterMapConfig.clustermapDeleteDataFromDatanodeConfigInPropertyStoreCleanUpTask &&
+        clusterMapConfig.clusterMapDataNodeConfigSourceType == DataNodeConfigSourceType.PROPERTY_STORE) {
+        synchronized (helixAdministrationLock) {
+          // Get the DataNodeConfig for the instance from the ZK
+          DataNodeConfig dataNodeConfig = dataNodeConfigSource.get(instance);
+          if (dataNodeConfig == null) { return;}
+
+          /* We will first try to remove data from local DataNodeConfig object
+             and then set it back in ZK based on configChanged flag
+          */
+          boolean configChanged;
+          // Remove all sealed, stopped, partially sealed and disabled replicas from DataNodeConfig
+          configChanged = TaskUtils.removeIfPresent(dataNodeConfig.getSealedReplicas());
+          configChanged = configChanged || TaskUtils.removeIfPresent(dataNodeConfig.getStoppedReplicas());
+          configChanged = configChanged || TaskUtils.removeIfPresent(dataNodeConfig.getPartiallySealedReplicas());
+          configChanged = configChanged || TaskUtils.removeIfPresent(dataNodeConfig.getDisabledReplicas());
+          Map<String, DataNodeConfig.DiskConfig> diskConfigs = dataNodeConfig.getDiskConfigs();
+
+          // Remove all replicas for each disk in the DataNodeConfig
+          for (DataNodeConfig.DiskConfig diskConfig : diskConfigs.values()) {
+            configChanged = configChanged || TaskUtils.removeIfPresent(diskConfig.getReplicaConfigs());
```

Comment:
java's logical operators will circuit break, which means that if configChanged is true, it will not execute `TaskUtils.removeIfPresent` method. So this is wrong. We should do this instead
```
configChanged = TaskUtils.removeIfPresent(diskConfig.getReplicaConfigs()) || configChanged;
```

Reply:
Thanks a lot for pointing this out. I have addressed the fix in https://github.com/linkedin/ambry/pull/2900